### PR TITLE
Pass Trainer to Models

### PR DESCRIPTION
**Pull request by blisc**

### CODE REVIEW

File: nemo/collections/asr/models/ctc_models.py [hunk 1]
```diff
@@ -44,8 +44,8 @@ class EncDecCTCModelConfig(ModelPTConfig):
 class EncDecCTCModel(ASRModel):
     """Encoder decoder CTC-based models."""
 
-    def __init__(self, cfg: EncDecCTCModelConfig):
-        super().__init__(cfg=cfg)
+    def __init__(self, cfg: EncDecCTCModelConfig, trainer=None):
```
Comment (blisc) [inline nemo/collections/asr/models/ctc_models.py:47]: Changed the init of EncDecCTCModel. I don't know if we want this since we wanted all models to be initialized with a DictConfig object. But I cannot figure out how to pass a python class with Hydra

File: nemo/core/classes/modelPT.py [hunk 2]
```diff
@@ -112,22 +113,22 @@ def setup_optimization(self, optim_config: Optional[Union[DictConfig, Dict]] = N
...
+        if 'sched' in optim_config and self._trainer is not None:
+            if not isinstance(self._trainer.accumulate_grad_batches, int):
+                raise ValueError("We do not currently support gradient acculumation that is not an integer.")
+            if self._trainer.max_steps is None:
+                if self._trainer.gpus == 0:
                     # training on CPU
-                    iters_per_batch = self._cfg.pl.trainer.max_epochs / float(
-                        self._cfg.pl.trainer.num_nodes * self._cfg.pl.trainer.accumulate_grad_batches
+                    iters_per_batch = self._trainer.max_epochs / float(
+                        self._trainer.num_nodes * self._trainer.accumulate_grad_batches
```
Comment (blisc) [inline nemo/core/classes/modelPT.py:123]: Note that now we can just grab the relevant information from the lightning trainer object as opposed to interpreting command line arugments

File: nemo/collections/asr/models/ctc_models.py (hunk 1 above)
Comment (okuchaiev) [reply nemo/collections/asr/models/ctc_models.py:47]: this was in my original draft, but I've reverted it back based on review feedback :) 
Do you really need to pass trainer instance? Or this https://github.com/NVIDIA/NeMo/blob/candidate/examples/asr/speech_to_text.py#L69 is fine?

Comment (blisc) [reply nemo/collections/asr/models/ctc_models.py:47]: PTL accepts all of the following for --gpu
```
# int: train on 2 gpus
trainer = Trainer(gpus=2)
# list: train on GPUs 1, 4 (by bus ordering)
trainer = Trainer(gpus=[1, 4])
trainer = Trainer(gpus='1, 4') # equivalent
# -1: train on all gpus
trainer = Trainer(gpus=-1)
trainer = Trainer(gpus='-1') # equivalent
# combine with num_nodes to train on multiple GPUs across nodes
# uses 8 gpus in total
trainer = Trainer(gpus=2, num_nodes=4)
``
I assume we want to support all of that. Now do we want to copy their initialize function and parse all of that, or do we just want to call trainer.init() and let them parse it and use trainer.num_gpus?

Comment (titu1994) [reply nemo/collections/asr/models/ctc_models.py:47]: I suppose its fine to pass in Trainer like this as it defaults to None anyway. we definitely dont want to cut-paste their logic, in case they change in the future (quite likely)

File: nemo/core/classes/modelPT.py [hunk 3]
```diff
@@ -112,22 +113,22 @@ def setup_optimization(self, optim_config: Optional[Union[DictConfig, Dict]] = N
...
         """
         # Setup optimizer and scheduler
-        if 'sched' in optim_config and self._cfg is not None and 'trainer' in self._cfg.pl:
-            if self._cfg.pl.trainer.max_steps is None:
-                if self._cfg.pl.trainer.gpus == 0:
+        if 'sched' in optim_config and self._trainer is not None:
+            if not isinstance(self._trainer.accumulate_grad_batches, int):
+                raise ValueError("We do not currently support gradient acculumation that is not an integer.")
+            if self._trainer.max_steps is None:
+                if self._trainer.gpus == 0:
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:120]: Do we need to check if `self._trainer.gpus` is an int, or a list prior to this? Or are we guaranteed an int by this point?

File: nemo/collections/asr/models/ctc_models.py (hunk 1 above)
Comment (tkornuta-nvidia) [reply nemo/collections/asr/models/ctc_models.py:47]: why do we need this "parsing"? cannot we can simply call:
```
python my_script.py trainer. gpus=2
```
?

Comment (blisc) [reply nemo/collections/asr/models/ctc_models.py:47]: @tkornuta-nvidia, how do I specify that I want to run on the 2nd gpu only? Or on the 3rd and 4th gpu only? How do we support `Trainer(gpus=[1, 4])`?

Comment (tkornuta-nvidia) [reply nemo/collections/asr/models/ctc_models.py:47]: If you are asking about Union, I am investigating it - but seems that OmegaConf doesn't support them...
